merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts

```diff
@@ -226,13 +226,21 @@ export class Windows {
    * @param fromStart duration to add from the start of each true segment
    * @param fromEnd duration to add from the end of each true segment
    */
-  public shiftBy(fromStart: AST.Duration, fromEnd: AST.Duration) : Windows {
-    return new Windows({
-      kind: AST.NodeKind.WindowsExpressionShiftBy,
-      windowExpression: this.__astNode,
-      fromStart,
-      fromEnd
-    })
+  public shiftBy(fromStart: AST.Duration, fromEnd?: AST.Duration | undefined) : Windows {
+    if (fromEnd === undefined) {
+      return new Windows({
+        kind: AST.NodeKind.ProfileExpressionShiftBy,
+        expression: this.__astNode,
+        duration: fromStart
+      });
+    } else {
+      return new Windows({
+        kind: AST.NodeKind.WindowsExpressionShiftBy,
+        windowExpression: this.__astNode,
+        fromStart,
+        fromEnd
+      })
+    }
```

Comment:
Wrapping my head around this... to rationalize to myself why both versions of this method are on the `Windows` class: profile expressions are a special case of windows expressions, so both of these operations make sense for profiles. Am I close?

As far as the semantics of the operations - `shiftBy(singleDuration)` takes the entire profile and moves it over by `singleDuration`. `shiftBy(deltaStart, deltaEnd)` takes every window individually and adjusts its start and end time accordingly. Does that track?

I'm wondering if these two operations are similar enough to share a name, or are they different enough such that this would cause confusion? I can be convinced either way here

Reply:
Perhaps a slightly different question would help me get a better feel for this - is `shiftBy(x)` equivalent to `shiftBy(x, x)`, or is there a subtle difference?

Reply:
Musing further - one could frame this PR as generalizing the `shiftBy` operation beyond boolean profiles. The part that's tricky is what to do if the operation causes two intervals to overlap - we'd need a way to collapse that multi-valued profile back to a single-valued profile. `Windows.shiftBy` effectively chooses "OR" as the merge operation. By restricting `shiftBy` on real and discrete profiles to `shiftBy(x, x)`, we dodge that issue, because no overlaps can happen. We also dodge tricky questions like what does it mean to stretch out a linear profile - does it reach a higher final value, or does it change its slope?

Reply:
yes, `windows.shiftBy(x)` is equivalent to `windows.shiftBy(x, x)`. I should add that to the doc comment.

> if the operation causes two intervals to overlap - we'd need a way to collapse that multi-valued profile back to a single-valued profile. *Windows.shiftBy effectively chooses "OR" as the merge operation.*

Not quite. `windows.shiftBy(fromStart, fromEnd)` can be thought of in two ways:
1. shift each true interval's bounds by `fromStart` and `fromEnd`, and apply the *converse* to each false interval (shift the bounds by `fromEnd` and `fromStart`.
2. or think of it as `windows.shiftBy(shiftRising, shiftFalling)` shifting the rising and falling edges. I've come to prefer this interpretation but it doesn't really matter.

The important thing is that both of those interpretations is that for any given segment/edge, the shift depends on whether it is true or false (or rising or falling). This makes it very difficult to extend to any other datatype: for real profiles, the shift would depend on whether the value/edge is....what? Also, the above interpretations don't produce an intermediate multi-valued boolean profile.

If we instead applied a uniform shift to all windows segments, we could get some undesired gaps:

```
input:
(-inf, 0) -> true
[0, inf) -> false

shiftBy(1, 0):
(-inf, 0) -> true
[1, inf) -> false
```

Reply:
Awesome, thanks for the explanation. The rising/falling edges mental model helps a lot